src/treelearner/serial_tree_learner.cpp

```diff
@@ -730,7 +727,7 @@ void SerialTreeLearner::ComputeBestSplitForFeature(
         leaf_splits->leaf_index(), config_->monotone_penalty);
     new_split.gain *= penalty;
   }
-  if (new_split > *best_split) {
+  if (new_split > *best_split && is_feature_used) {
```

Comment:
```suggestion
  // it is needed to filter the features after above code. 
  // Otherwise, the `is_splittable` in `FeatureHistogram` will be wrong, and cause some features being accidently filtered in the later nodes.
  if (new_split > *best_split && is_feature_used) {
```